lib/tests/test_working_copy.rs

```diff
@@ -669,6 +683,81 @@ fn test_dotgit_ignored(use_git: bool) {
     locked_wc.discard();
 }
 
+#[test]
+fn test_gitsubmodule() {
+    // Tests that git submodules are ignored.
+
+    let settings = testutils::user_settings();
+    let mut test_workspace = TestWorkspace::init(&settings, true);
+    let repo = &test_workspace.repo;
+    let store = repo.store().clone();
+    let workspace_root = test_workspace.workspace.workspace_root().clone();
+
+    let mut tree_builder = store.tree_builder(store.empty_tree_id().clone());
+
+    let added_path = RepoPath::from_internal_string("added");
+    let submodule_path = RepoPath::from_internal_string("submodule");
+    let added_submodule_path = RepoPath::from_internal_string("submodule/added");
+
+    tree_builder.set(
+        added_path.clone(),
+        TreeValue::File {
+            id: testutils::write_file(&repo.store(), &added_path, "added\n"),
+            executable: false,
+        },
+    );
+
+    let mut tx = repo.start_transaction(&settings, "create submodule commit");
+    let submodule_id = create_random_commit(&settings, repo)
+        .write_to_repo(tx.mut_repo())
+        .id()
+        .clone();
+    tx.commit();
+
+    tree_builder.set(
+        submodule_path.clone(),
+        TreeValue::GitSubmodule(submodule_id),
+    );
+
+    std::fs::create_dir(workspace_root.join("submodule")).unwrap();
+
+    testutils::write_working_copy_file(
+        &workspace_root,
+        &added_submodule_path,
+        "i am a file in a submodule\n",
+    );
+
+    let tree_id = tree_builder.write_tree();
+
+    let tree = store.get_tree(&RepoPath::root(), &tree_id).unwrap();
+
+    let wc = test_workspace.workspace.working_copy_mut();
+    wc.check_out(repo.op_id().clone(), None, &tree).unwrap();
+
+    // Check that the working copy is clean.
+    let mut locked_wc = wc.start_mutation();
+    let new_tree_id = locked_wc.snapshot(GitIgnoreFile::empty()).unwrap();
+    locked_wc.discard();
+    assert_eq!(new_tree_id, tree_id);
+
+    // Check that the files in the submodule are not tracked
+    let tree1 = repo
+        .store()
+        .get_tree(&RepoPath::root(), &new_tree_id)
+        .unwrap();
```

Comment:
Since you checked before that `new_tree_id` is the same as `tree_id`, there's no need to re-read the tree here, and thus no need to check the contents of the tree. So I think line 741 is enough to "Check that the files in the submodule are not tracked"